### PR TITLE
Explicitly associate fetched build scan data with the run num

### PR DESCRIPTION
**Pull request by jthurne**
The fix for "Script does not list second build twice when first build fails"
accidentally broke offline mode because parsing fetched Build Scan data
depended on the Build Scan ID to re-establish the association between the
fetched Build Scan data and the run num (build number). Unfortunately, there
are no Build Scan IDs when the scripts are executed in offline mode.

This commit fixes things by explicitly passing the run num with the Build Scan
URLs when invoking the Java utilities to fetch Build Scan data or to read Build
Scan data from the local Build Scan dumps.

Interestingly, processing within fetch-build-scan-data-cmdline-tool was already
oriented around the run num. It used the run num in informational log messages
to let the user know which build it was fetching data for. This made it
relatively easy to pass run num through the processing chain.

### CODE REVIEW

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java [hunk 1]
```diff
@@ -58,6 +58,7 @@ public static class ExitCode {
 
     @Override
     public Integer call() throws Exception {
+        List<Build> builds = buildsFrom(runNumsAndBuildScanUrls);
```
Comment (etiennestuder) [inline components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java:61]: new line after

Comment (etiennestuder) [reply components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java:61]: or move to right above the `customValueKeys` declaration

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java [hunk 2]
```diff
@@ -66,11 +67,10 @@ public Integer call() throws Exception {
...
 
         logStartFetchingBuildScans();
-        List<BuildValidationData> buildValidationData = new ArrayList<>();
-        for (int i = 0; i < buildScanUrls.size(); i++) {
-            BuildValidationData validationData = fetchBuildScanData(i, buildScanUrls.get(i), customValueKeys);
-            buildValidationData.add(validationData);
-        }
+        List<BuildValidationData> buildValidationData = builds.stream()
+            .map(build -> fetchBuildScanData(build.runNum(), build.buildScanUrl(), customValueKeys))
+            .collect(Collectors.toList());
```
Comment (etiennestuder) [inline components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java:72]: Nice.

Comment (etiennestuder) [reply components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java:72]: I think I would extract this block into a separate method that returns `List<BuildValidationData> buildValidationData`. The reason is to have the same level of abstraction in this section.

And the same method extraction further down with `buildValidationData.forEach(this::printRow);`

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java [hunk 3]
```diff
@@ -148,6 +149,22 @@ private String buildScanIndexToOrdinal(int i) {
...
+        return runNumsAndBuildScanUrls.stream().map(this::buildFrom)
+            .collect(Collectors.toList());
+    }
+
+    private Build buildFrom(String runNumAndBuildScanUrl) {
+        String[] parts = runNumAndBuildScanUrl.split(",");
+        String runNum = parts[0];
+        String buildScanUrl = parts[1];
+        try {
+            return new Build(Integer.parseInt(runNum), new URL(buildScanUrl));
```
Comment (etiennestuder) [inline components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java:162]: I would introduce a factory method on the `Build` class that takes two strings.

Comment (etiennestuder) [reply components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java:162]: Or possibly even takes the entire comma-separated String.

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java [hunk 4]
```diff
@@ -206,28 +223,27 @@ private void logStartFetchingBuildScan(int index) {
         }
     }
 
-    private String fetchingMessageFor(int index) {
-        if (buildScanUrls.size() <= 10) {
-            return String.format("Fetching build scan data for the %s build", buildScanIndexToOrdinal(index));
+    private String fetchingMessageFor(int runNum) {
+        if (runNumsAndBuildScanUrls.size() <= 10) {
```
Comment (etiennestuder) [inline components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java:227]: I would pass in `runNumsAndBuildScanUrls` to this method. IMO, the params passed originally get processed in call() and should not be accessed as instance fields anymore from then on.

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java [hunk 5]
```diff
@@ -132,7 +133,7 @@ private void printFailedRequest(FailedRequestException e) {
         }
     }
 
-    private String buildScanIndexToOrdinal(int i) {
+    private String runNumToOrdinal(int i) {
```
Comment (etiennestuder) [inline components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java:136]: Make static. Move to the end of this class.

**Review by etiennestuder — CHANGES_REQUESTED** [review]

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java (hunk 5 above)
Comment (etiennestuder) [inline components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java:136]: I would simplify the implementation to return String.valueOf(i)+ "th" as the default for all values >=4.

You can then also simplify `fetchingMessageFor()` by removing the size check.

**Review by etiennestuder — CHANGES_REQUESTED** [review]

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/model/Build.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+import java.net.URL;
+
+public class Build {
+    private final int runNum;
+
+    private final URL buildScanUrl;
+
+    public Build(int runNum, URL buildScanUrl) {
+        this.buildScanUrl = buildScanUrl;
+        this.runNum = runNum;
```
[the remark this answers is not in the document]
Comment (jthurne) [reply components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/model/Build.java]: ✅ Done

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java (hunk 1 above)
Comment (jthurne) [reply components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java]: ✅ Done

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java (hunk 2 above)
Comment (jthurne) [reply components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java]: ✅ Done

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java (hunk 3 above)
Comment (jthurne) [reply components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java]: ✅ Done

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java (hunk 4 above)
Comment (jthurne) [reply components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java]: ✅ Eliminating the size check eliminates the need to access runNumsAndBuildScanUrls at all.  In fact, eliminating that size check reduced this method to a single line, which only called in one place, so I inlined it.

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java (hunk 5 above)
Comment (jthurne) [reply components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java]: ✅ Done

Comment (jthurne) [reply components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java]: ✅ Done

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java [hunk 7]
```diff
@@ -242,8 +213,8 @@ private void logFetchResults(List<BuildValidationData> buildValidationData, Cust
         }
     }
 
-    private void logFetchResultFor(int index, String property, String customValueKey, boolean found) {
-        String ordinal = runNumToOrdinal(index);
+    private void logFetchResultFor(int runNum, String property, String customValueKey, boolean found) {
+        String ordinal = toOrdinal(runNum);
```
Comment (etiennestuder) [inline components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java:217]: I would inline this local variable with 1 usage.

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java [hunk 8]
```diff
@@ -252,16 +223,29 @@ public void printHeader() {
         System.out.println(String.join(",", labels));
     }
 
+    private void printBuildValidationData(List<BuildValidationData> buildValidationData) {
```
Comment (etiennestuder) [inline components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java:226]: Maybe use 
- `printBuildValidationDataItems` and `printBuildValidationDataItem` or
- `printAllBuildValidationData` and `printSingleBuildValidationData`

**Review by etiennestuder — CHANGES_REQUESTED** [review]

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java [hunk 9]
```diff
@@ -58,45 +59,50 @@ public static class ExitCode {
...
-        List<Build> builds = buildsFrom(runNumsAndBuildScanUrls);
         // Use System.err for logging since we're going to write out the CSV to System.out
         logger = new ConsoleLogger(System.err, colorScheme, debug);
 
         networkSettingsFile.ifPresent(settingsFile -> NetworkSettingsConfigurator.configureNetworkSettings(settingsFile, logger));
 
+        List<NumberedBuildScan> buildScans = parseBuildScans(runNumsAndBuildScanUrls);
         CustomValueNames customValueKeys = loadCustomValueKeys(customValueMappingFile);
 
         logStartFetchingBuildScans();
```
[the remark this answers is not in the document]
Comment (jthurne) [reply components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java]: ✅ Done

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/model/NumberedBuildScan.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    public static NumberedBuildScan parse(String runNumAndBuildScanUrl) {
+        String[] parts = runNumAndBuildScanUrl.split(",");
+        String runNum = parts[0];
+        String buildScanUrl = parts[1];
+        try {
+            return new NumberedBuildScan(Integer.parseInt(runNum), new URL(buildScanUrl));
+        } catch (MalformedURLException e) {
+            throw new BadBuildScanUrlException(buildScanUrl, e);
```
[the remark this answers is not in the document]
Comment (jthurne) [reply components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/model/NumberedBuildScan.java]: ✅ Removed

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java [hunk 11]
```diff
@@ -252,16 +223,29 @@ public void printHeader() {
...
+    }
+
     private void printRow(BuildValidationData buildValidationData) {
         List<String> values = Fields.ordered().map(f -> f.value.apply(buildValidationData)).collect(Collectors.toList());
         System.out.println(String.join(",", values));
     }
 
-    private CustomValueNames loadCustomValueKeys(Optional<Path> customValueMappingFile) throws IOException {
+    private static CustomValueNames loadCustomValueKeys(Optional<Path> customValueMappingFile) throws IOException {
         if (customValueMappingFile.isPresent()) {
```
[the remark this answers is not in the document]
Comment (jthurne) [reply components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java]: ✅ Done

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java (hunk 7 above)
Comment (jthurne) [reply components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java]: ✅ Done

File: components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java (hunk 8 above)
Comment (jthurne) [reply components/fetch-build-scan-data-cmdline-tool/src/main/java/com/gradle/enterprise/cli/FetchBuildValidationDataCommand.java:198]: ✅ Done

**Review by etiennestuder — APPROVED** [review]

**Pull request merged**